### PR TITLE
Add playback options to settings

### LINKED ISSUE #2765: Make frames per second of playback a setting
## 🚀 Feature
I'm doing some stuff right now where it would be very nice to have the fps (and direction/ playmode) of the playback be settings so I don't have to change them every time I re-open napari - i.e. things in:

Screen Shot 2021-05-22 at 9 25 47 PM

Specified here https://github.com/napari/napari/blob/bdb2887baac00d01f89426951b82cb523af1371a/napari/_qt/widgets/qt_dims_slider.py#L73-L76

### PR DESCRIPTION
**Pull request by ppwadhwa**
# Description

This PR adds some of the playback options (fps, playback mode, and direction) into settings to persist from session from session.

Closes #2765 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by ppwadhwa** [comment]
I moved the _constants.py file because I was getting a circular imports error when I wanted to use it in _defaults.py. It seems to work fine having it in the new location...

**Comment by ppwadhwa** [comment]
I see some of the tests failing and some of them, I don't know why-- see  ubuntu-latest 3.8.

On the test that failed in napari/_qt/widgets/_tests/test_qt_dims.py, it seems to be a timing issue.  I'm not sure how to fix it.  When I look at it locally, it will sometimes fail and sometimes pass.

**Comment by tlambert03** [comment]
oh, that was my least favorite test... but I've learned a thing or two since writing that crappy test... so maybe can improve now :)  will look

**Comment by ppwadhwa** [comment]
@tlambert03 haha---I've stumbled upon the **least** favorite??  Thanks for looking at it.  :)

**Comment by tlambert03** [comment]
a thought here... while I like having some general persistence for these settings, i'm not so sure that I'd personally want the changes I make to the widgets in the play button to stick across sessions.  (loop mode and play direction in particular seem like something I generally want _not_ to stick).

@sofroniewn, was that type of persistence important for what you were wanting here?  would it be enough to just recover the FPS?

**Comment by sofroniewn** [comment]
> @sofroniewn, was that type of persistence important for what you were wanting here? would it be enough to just recover the FPS?

FPS would be enough for me, i just found myself constantly opening napari and changing it every time i played a movie

>  i'm not so sure that I'd personally want the changes I make to the widgets in the play button to stick across sessions. (loop mode and play direction in particular seem like something I generally want not to stick).

that seems reasonable, so i guess behavior would be that on load value gets propagated from settings into the widget, but that if value changed in widget it wouldn't change in settings. If you wanted to permanently change in settings you'd have to go into settings and change it, which would then update the value in the widget.

That seems very reasonable to me and would certainly meet my needs. @ppwadhwa is that something easy for you to implement in the current system?

I'm also ok with just doing FPS, and maybe loop mode (i could see how someone might always prefer one type of mode). Direction seems a little funny to have as always going backwards!!!

Is this clear?

**Comment by ppwadhwa** [comment]
@sofroniewn Only update FPS and loop mode in settings when changed in the widget?  Still keep direction as a setting, but don't update that value when the user clicks that button?

I can make these changes.   :)

**Comment by sofroniewn** [comment]
> @sofroniewn Only update FPS and loop mode in settings when changed in the widget? Still keep direction as a setting, but don't update that value when the user clicks that button?

No I meant to say keep FPS and loop mode as settings but don't change them if changed only in play mode widget, and drop direction as setting entirely.

Is that what you want too @tlambert03?

My bad for confusion

**Comment by ppwadhwa** [comment]
@sofroniewn got it!  thanks for clarifying.  I will make these changes.

**Review by goanpeca — APPROVED** [review]
Thanks for working on this, looks good :)

File: napari/_qt/widgets/qt_dims_slider.py [hunk 1]
```diff
@@ -70,10 +71,21 @@ def __init__(self, parent: QWidget, axis: int):
...
+        settings = get_settings()
+
+        # For whatever reason, using settings.application.playback_mode by itself
+        # does not work. But setting the mode this way seems to work.
+        if settings.application.playback_mode == 'loop':
+            loop_mode = LoopMode.LOOP
+        elif settings.application.playback_mode == 'back_and_forth':
+            loop_mode = LoopMode.BACK_AND_FORTH
+        else:
+            loop_mode = LoopMode.ONCE
```
Comment (tlambert03) [inline napari/_qt/widgets/qt_dims_slider.py]: yeah, this bugs me too.  the reason is that we have the following config set for our settings (and our evented model).

```py
class Config:
    use_enum_values = True
```

there was a desire earlier on to keep things as plan strings for the outside world... hence the setting... but I think we're reconsidering that.  For now though, you can simplify this as follows (i.e. you can call the enum class with the string value)

```suggestion
        LoopMode(settings.application.playback_mode)
```

**Review by tlambert03 — COMMENTED** [review]
Hi @ppwadhwa, I'm a bit confused now on how to persist the setting ... is this supposed to appear in the preferences?

Untitled

**Comment by goanpeca** [comment]
> I'm a bit confused now on how to persist the setting ... is this supposed to appear in the preferences?

Nope, it will store whatever the user sets on the dialog.

**Comment by tlambert03** [comment]
> Nope, it will store whatever the user sets on the dialog.

ok... then it's not working for me :joy:

**Comment by ppwadhwa** [comment]
I had it store them,  then I thought Nick said not to persist them this way
(see above), but the user would go directly to set them in the settings
file.  Maybe he meant to have them in the preferences dialog.  I will add
something there.   Sorry about that.

**Comment by tlambert03** [comment]
> then I thought Nick said not to persist them this way (see above), but the user would go directly to set them in the settings file. 

my take on nicks comment

> keep FPS and loop mode as settings but don't change them if changed only in play mode widget, and drop direction as setting entirely.

was that the emphasis was on the "if changed **only** in play widget"... but not that it shouldn't be included at all in the preferences dialog. is that correct @sofroniewn?

**Comment by ppwadhwa** [comment]
Yeah, I agree having re-read it now.  But, just one more time for myself to
make sure I am still not misinterpreting---   add these options to
preferences dialog.  Save from there, don't save from widget itself.  If
you change in preferences dialog, the widget should update.

Sorry about my brain being in a fog!

**Comment by tlambert03** [comment]
No worries!  Yeah that’s my take on it.  Will let Nick confirm

**Comment by ppwadhwa** [comment]
@tlambert03  @sofroniewn @goanpeca Take a look!  Let me know if this is right.  :)


https://user-images.githubusercontent.com/54282105/124227611-9868f480-dad0-11eb-89f7-a024e22b739c.mov

**Comment by sofroniewn** [comment]
Just confirmed this works great for me!

**Review by sofroniewn — APPROVED** [review]

**Comment by sofroniewn** [comment]
This now has a merge conflict @ppwadhwa - can you give it a fix, then i'll be happy to merge. thanks!!

**Comment by ppwadhwa** [comment]
@sofroniewn   I have rebased to master.  When checking that everything was working, I noticed that the settings were not being saved.  So, if I re-opened napari, my new values were not set, but only loading the defaults.  I went to the master branch and tried out a few settings and found that others were not saving either (I tried theme and shortcuts).    Is it just me or is this a problem for others?

**Comment by sofroniewn** [comment]
> @sofroniewn I have rebased to master. When checking that everything was working, I noticed that the settings were not being saved. So, if I re-opened napari, my new values were not set, but only loading the defaults. I went to the master branch and tried out a few settings and found that others were not saving either (I tried theme and shortcuts). Is it just me or is this a problem for others?

Broken for me too @ppwadhwa!! Looks like we have a regression! Not sure where it crept in, my guess is recent though

**Comment by sofroniewn** [comment]
@ppwadhwa looks like regression came in https://github.com/napari/napari/pull/2998, would be great if you could make a PR to fix, while preserving behaviour of #2998 too, thanks!!

**Comment by ppwadhwa** [comment]
@sofroniewn Just made a PR that should fix the problem --> #3041.  Please have a look when you get a moment!

**Comment by sofroniewn** [comment]
Ok, this like it's now ready to go, thanks @ppwadhwa!!

**Pull request merged**